### ISSUE TITLE
[Blazor] Hot Reload not working properly when using RenderModeAuto in .NET 8 RC1

### ISSUE THREAD

**Issue by bub-bl** [dotnet/aspnetcore#51236]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I have created a new **Blazor Web App** project from Visual Studio 17.8.0 Preview 2 with WebAssembly and Server interactivity. But the hot reload doesn't work properly. When i add some code in a razor file, i need to refresh the browser manually to see changes

## Client project

- When i add some code in any page or component, the component is refreshed automaticaly in the browser, only if the render mode is set. If the component have no interactivity, hot reload doesn't update the component in the browser.

## Server project

- When i add some code in any page or component, hot reload refresh the browser only if a render mode is set but doesn't work when the render mode is set to RenderModeAuto or if the component have no interactivity.

### Expected Behavior

Any page or component with any render mode need to be refreshed automaticaly when the hot reload is triggered.

### Steps To Reproduce

- Create new **Blazor Web App** project with WebAssembly and Server interactivity.
- Update any page or component inside of the Client project -> The browser
is not refreshed automaticaly when the component have no interactivity, you need to refresh the browser manually to see changes.

- Update any page or component at inside of the Server project -> The browser is not updated automaticaly when hot reload is triggered with the RenderModeAuto attribute, you need to refresh the browser manually to see changes.

### Exceptions (if any)

![image](https://github.com/dotnet/aspnetcore/assets/66354328/fe64cc5d-ab37-4fee-b2ee-3cd961793df0)

### .NET Version

8.0.100-rc.1.23463.5

### Anything else?

C:\Users\bubbl>dotnet --info
SDK .NET :
 Version:   8.0.100-rc.1.23463.5
 Commit:    e7f4de8816

Environnement d'exécution :
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100-rc.1.23463.5\

.NET workloads installed:
 [maui-android]
   Source de l’installation: SDK 8.0.100-rc.1
   Version de manifeste:    8.0.0-rc.1.9171/8.0.100-rc.1
   Chemin d'accès au Manifeste:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.1\microsoft.net.sdk.maui\8.0.0-rc.1.9171\WorkloadManifest.json
   Type d'installation:              Msi

 [maui-desktop]
   Source de l’installation: SDK 8.0.100-rc.1
   Version de manifeste:    8.0.0-rc.1.9171/8.0.100-rc.1
   Chemin d'accès au Manifeste:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.1\microsoft.net.sdk.maui\8.0.0-rc.1.9171\WorkloadManifest.json
   Type d'installation:              Msi

 [maui-ios]
   Source de l’installation: SDK 8.0.100-rc.1
   Version de manifeste:    8.0.0-rc.1.9171/8.0.100-rc.1
   Chemin d'accès au Manifeste:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.1\microsoft.net.sdk.maui\8.0.0-rc.1.9171\WorkloadManifest.json
   Type d'installation:              Msi

 [maui-windows]
   Source de l’installation: SDK 8.0.100-rc.1, VS 17.8.34112.27
   Version de manifeste:    8.0.0-rc.1.9171/8.0.100-rc.1
   Chemin d'accès au Manifeste:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.1\microsoft.net.sdk.maui\8.0.0-rc.1.9171\WorkloadManifest.json
   Type d'installation:              Msi

 [wasi-experimental]
   Source de l’installation: SDK 8.0.100-rc.1
   Version de manifeste:    8.0.0-rc.1.23419.4/8.0.100-rc.1
   Chemin d'accès au Manifeste:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.1\microsoft.net.workload.mono.toolchain.current\8.0.0-rc.1.23419.4\WorkloadManifest.json
   Type d'installation:              Msi

 [maccatalyst]
   Source de l’installation: VS 17.8.34112.27
   Version de manifeste:    16.4.8825-net8-rc1/8.0.100-rc.1
   Chemin d'accès au Manifeste:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.1\microsoft.net.sdk.maccatalyst\16.4.8825-net8-rc1\WorkloadManifest.json
   Type d'installation:              Msi

 [ios]
   Source de l’installation: VS 17.8.34112.27
   Version de manifeste:    16.4.8825-net8-rc1/8.0.100-rc.1
   Chemin d'accès au Manifeste:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.1\microsoft.net.sdk.ios\16.4.8825-net8-rc1\WorkloadManifest.json
   Type d'installation:              Msi

 [android]
   Source de l’installation: VS 17.8.34112.27
   Version de manifeste:    34.0.0-rc.1.432/8.0.100-rc.1
   Chemin d'accès au Manifeste:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.1\microsoft.net.sdk.android\34.0.0-rc.1.432\WorkloadManifest.json
   Type d'installation:              Msi

Host:
  Version:      8.0.0-rc.1.23419.4
  Architecture: x64
  Commit:       92959931a3
  RID:          win-x64

.NET SDKs installed:
  7.0.306 [C:\Program Files\dotnet\sdk]
  8.0.100-rc.1.23463.5 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 7.0.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-rc.1.23421.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 7.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-rc.1.23419.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 7.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0-rc.1.23420.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. Can you please try out the latest RC2 release which will be shipped later today and let us know if you still experience those issues?

**Comment by bub-bl** [comment]
@mkArtakMSFT I'll try again when RC2 is available

**Comment by bub-bl** [comment]
@mkArtakMSFT I have tryed the new .NET 8 RC2 update, i have updated all dependencies in both projects. All works perfectly, when i add some code into the Counter page inside of the Client project, hot reload is restarting the application every time when the render mode is set to RenderModeInteractiveAuto. If i use an another render mode, the hot reload just update the browser correctly. 

I think the problem is related to the C# code when we use the @code directive or when we use any code inside the razor file.

**Comment by Markz878** [comment]
At least in WASM render mode the hot reload is not working currently, just changing element classes is enough to break hot reload...

**Comment by bub-bl** [comment]
> At least in WASM render mode the hot reload is not working currently, just changing element classes is enough to break hot reload...

I have checked page and components with the rendermode RenderModeInteractiveWebAssembly, i dont have any problem. 

Are you using .NET 8 RC1 or RC2 ?
Have you added the following code inside of the Server Project in the Program.cs file ?
```csharp
builder.Services.AddRazorComponents()
    .AddInteractiveServerComponents()
    .AddInteractiveWebAssemblyComponents();
    
app.MapRazorComponents<App>()
    .AddAdditionalAssemblies(typeof(BlazorApp1.Client._Imports).Assembly)
    .AddInteractiveServerRenderMode()
    .AddInteractiveWebAssemblyRenderMode();
```

**Comment by Markz878** [comment]
I'm using a fresh RC2 template, so yes I have all of that. The issue only happens when creating a new RC2 blazor web app from Visual Studio, with the command line it works fine, pretty strange..

**Comment by mkArtakMSFT** [comment]
@Markz878 can you please clarify this please? Are you saying the very same project created from VS and CLI has different output?

> The issue only happens when creating a new RC2 blazor web app from Visual Studio, with the command line it works fine, pretty strange..

@bub-bl we are aware of some WASM specific hot reload issues which we're tracking already as part of https://github.com/dotnet/aspnetcore/issues/50765

I think we will need to fix that first and if this is not resolved, then we'll look into this separately.

**Comment by danroth27** [comment]
> when i add some code into the Counter page inside of the Client project, hot reload is restarting the application sometimes when the render mode is set to RenderModeInteractiveAuto. If i use an another render mode, the hot reload just update the browser correctly.

@bub-bl Could you please share with me the specific edits you attempt in the client project that trigger the unexpected unsupported edit?

**Comment by bub-bl** [comment]
@danroth27 I have tested a lot of things and the problem seam to appear only when i use RenderModeInteractiveWebAssembly or RenderModeInteractiveAuto. The application restart when any change is provided in razor file. This is not related to C# code inside the razor file.
This is also pretty random, that can be work nice, or restart when a changed is detected. Currently, i dont have the problem with the render mode set to WebAssembly, but i have this warning in the console when i do a change. (However, hot reload has updated correctly the browser correctly)

I dont have any interactivity when the render mode is set to WebAssembly.

![image](https://github.com/dotnet/aspnetcore/assets/66354328/b8f91a2d-6873-4c13-bc2d-db2467432c65)

The problem does not appear with no interactivity or when i use RenderModeInteractiveServer

**Comment by bub-bl** [comment]
I can clone my current project if you need to do some tests

**Comment by danroth27** [comment]
I tried creating a Blazor Web App (Auto, per-page) and then change the counter and I'm getting this BrowserConnectionException error:

![image](https://github.com/dotnet/aspnetcore/assets/1874516/9bb4d0fd-93b1-441d-b748-776c45364564)

I opened https://developercommunity.visualstudio.com/t/MicrosoftWebToolsProjectSystemHotRelo/10489615 in case this needs to be looked at on the VS tooling side.

**Comment by oliverw** [comment]
> I tried creating a Blazor Web App (Auto, per-page) and then change the counter and I'm getting this BrowserConnectionException error:
> 
> ![image](https://user-images.githubusercontent.com/1874516/274377984-9bb4d0fd-93b1-441d-b748-776c45364564.png)
> 
> I opened https://developercommunity.visualstudio.com/t/MicrosoftWebToolsProjectSystemHotRelo/10489615 in case this needs to be looked at on the VS tooling side.
> 
> I get the same error with just WebAssembly and per-page.

WASM Hot-reload appears to be completely broken in RC2. I'm getting the exact same error after changing a server side Blazor Page. (`Blazor Web App` project type). So at least for me hot-reload does not exist on either side.

**Comment by bub-bl** [comment]
@oliverw Have you tried with the `dotnet watch` command ?

**Comment by oliverw** [comment]
> @oliverw Have you tried with the `dotnet watch` command ?

Nope, only tried to run in the VS 2022 17.8 Debugger.

**Comment by rudiv** [comment]
Can confirm this with `dotnet watch`.

When using `RenderMode.InteractiveAuto`, the server changes are applied as you would expect and you see the flash for a second before WASM takes over, and the WASM project shows the old content.

For full reference this is just changing the title in the `<h1>` of Home.razor.

The output is as follows:

```
dotnet watch ⌚ File changed: ./.Client/Pages/Home.razor.
dotnet watch ⌚ No changes applied to client because they are not supported by the runtime.
dotnet watch 🔥 Hot reload of changes succeeded.
```

**Comment by bub-bl** [comment]
@rudiv Have you the same problem when you create the project with this following line ?
`dotnet new blazor -n BlazorAppX -int WebAssembly -ai`

Do not miss to add InteractiveServer function in Program.cs

**Comment by rudiv** [comment]
Now there is no "error", the output is as follows:

```
dotnet watch ⌚ File changed: ./.Client/Pages/Home.razor.
dotnet watch 🔥 Hot reload of changes succeeded.
```

However, the WASM still takes over and is the old version even after refresh. Specifically the title is changed, and after ~500ms it switches back to the version without the change.

A second change results in the rude edit warning after which, of course, the WASM is recompiled and shipped down.
```
dotnet watch ⌚ File changed: ./.Client/Pages/Home.razor.
  ❔ Do you want to restart your app - Yes (y) / No (n) / Always (a) / Never (v)?
  ❔ a
```

Third change same as the first, fourth rude edit, fifth same as first, sixth rude edit etc. It's almost as though it knows the changes are not applied and so wants to fully recompile (which works).

Again this is just changing the text within the `<h1>`.

My expected behaviour just for clarity would be that both the Server and WASM components are recompiling together, without this it seems as though development in Blazor WASM would be extremely difficult.

**Comment by oliverw** [comment]
This needs to be fixed before GA or development experience will be quite degraded.

**Comment by danroth27** [comment]
@oliverw the BrowerConnectionException issue I [mentioned](https://github.com/dotnet/aspnetcore/issues/51236#issuecomment-1758246165) above should be fixed soon in the next Visual Studio 17.8 preview update. We have a number of other hot reload fixes that are also coming as part of the upcoming .NET 8 release in November.  I have verified that hot reload works correctly both with the static server rendered components and the client components when using a Blazor Web App with Auto enabled. I tested this both in our latest internal VS builds and with `dotnet watch`. If you want to try a .NET 8 RTM build, they are available publicly from the https://github.com/dotnet/installer repo.

@mkArtakMSFT I think we're in good shape here, so I'm going to close this as fixed in .NET 8.
